### ISSUE TITLE
Get rid of deprecated componentWillReceiveProps

### ISSUE THREAD

**Issue by ctavan** [nayaabkhan/react-polyglot#38]
`componentWillReceiveProps` is [deprecated and will be removed in React 17](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops).

It's currently being used to update the internal node-polyglot instance when the `<I18n>` component's props change: https://github.com/nayaabkhan/react-polyglot/blob/db80a6374505944a6bae3bc150c14364863c176d/src/i18n.js#L22-L30

We need to find a future-proof way of updating the internal node-polyglot instance.

**Comment by ctavan** [comment]
From @soda0289 in https://github.com/nayaabkhan/react-polyglot/pull/35#issuecomment-543378648 :

> @ctavan @nayaabkhan
> 
> The issue with componentDidUpdate is that although we do update the polyglot instance it only happens after we have rendered.
> 
> Is there any reason why we cant just do the check inside render().
> 
> Something like
> 
> ```js
> if (this.polyglot.currentLocale  !== this.props.locale) {
>   this.polyglot.locale(this.props.locale);
> }
> ```
> 
> That way we update the instance then create a new bound function that would trigger a re render of the Context consumers.
> 
> I can work on this change as well since we have run into issues where changing locale wont trigger re render.
> 
> **EDIT: The workaround we use now is to pass in a key prop for `<I18n />` component that uses the locale.
> `<I18n key={locale} locale={locale} />`

**Comment by ctavan** [comment]
@soda0289 I believe your suggestion of updating the polyglot instance in the render-method should work well for `props.locale` since by comparing the string this operation can be made idempotent and efficient.

I'm not so sure about the messages though: When calling `this.polyglot.replace()` the messages object is copied internally by `node-polyglot` so we cannot (shallow) compare `props.messages` with some value inside the `node-polyglot` instance.

At a first glance I would assume that what we want is the `memoize-one` solution from 
https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization

What do you think?

**Comment by soda0289** [comment]
I think memoizing the input props is the correct solution.

If we use the new react hooks they offer useMemo() which would work well in this siutaion but we would have to convert the component class into a component function to use hooks.

```js
function I18n({children, locale, messages}) {
   const translate = React.useMemo(() => {
       this._polyglot.locale(this.props.locale)
       this._polyglot.replace(this.props.messages)

       return this._polyglot.t.bind(this._polyglot)
   }, [locale, messages]);

    return (
      <I18nContext.Provider value={translate}>
        {React.Children.only(children)}
      </I18nContext.Provider>
    )
  }
}
```

If we used memoize-one would it just be something like:

```js
  function render() {
    const getTranslate = memoizeOne(
      (locale, messages) => {
        this._polyglot.locale(this.props.locale)
        this._polyglot.replace(this.props.messages)

        return this._polyglot.t.bind(this._polyglot)
      }
    );
   const translate = getTranslate(locale, messages);

    return (
      <I18nContext.Provider value={translate}>
        {React.Children.only(children)}
      </I18nContext.Provider>
    );
  }
```


Would it be ok to do a shallow comparsion of messages object. For my use case I always load the messages from a json file so the object never changes. Which would allow for shallow or "===" comparison.

**Comment by ctavan** [comment]
I believe we should do a shallow comparison only (and probably document this). Anything else could result in serious performance problems.

**Comment by soda0289** [comment]
@ctavan @nayaabkhan Any preference between using the react hooks with function component or should we be using the class component and bring in the memozineOne dependency?

**Comment by ctavan** [comment]
I‘m personally a big fan of hooks and would favor it.

We just need to make sure to adjust the react peerDependency accordingly and release the new version of this library as 0.7.0

**Comment by ctavan** [comment]
Speaking of hooks: It would then of course also be nice to have a `const [t] = useTranslate();` hook instead of having to continue to use the HOC. 😉 Maybe as an additional API?

**Comment by soda0289** [comment]
@ctavan We added a hook already https://github.com/nayaabkhan/react-polyglot/pull/22

Much nicer API then HOCs. Especially with typescript typing.

**Comment by ctavan** [comment]
Whoops, of course! Sorry for missing out on that one!

**Comment by nayaabkhan** [comment]
Agreed, it is time to migrate toward using hooks internally and update our `peerDependency` version for `react` to `16.8.0`.

**Comment by nayaabkhan** [comment]
I think it would be fine to move our React `peerDependency` ahead to `16.8.0` and use the `useMemo` hook internally. In any case, people who aren't looking to migrate to a newer version of React can keep using the current version of `react-polyglot` without issues.

**Comment by ctavan** [comment]
Cool!

@soda0289 are you planning to make a pull request? Just so we don’t work both on it at the same time.

**Comment by soda0289** [comment]
@ctavan Yes I can work on this today or tommrow.

**Comment by puredazzle** [comment]
Hi! Any updates on this issue?

**Comment by nayaabkhan** [comment]
@puredazzle Hi! I will check the PR #39 that resolves this issue in the coming days.

**Comment by nayaabkhan** [comment]
Released `v0.7.0` 🎉
